tools/github/release/abstract.py

```diff
@@ -0,0 +1,313 @@
+import pathlib
+from abc import abstractmethod
+from functools import cached_property
+from typing import (
+    Any, AsyncGenerator, Awaitable, Coroutine, Dict, Iterable, Iterator, List, Optional, Pattern,
+    Set, Type, Union)
+
+import verboselogs  # type:ignore
+
+import packaging.version
+
+import aiohttp
+
+import gidgethub.abc
+
+from tools.base import abstract
+from tools.base.functional import async_property
+
+
+class AGithubReleaseAssets(metaclass=abstract.Abstraction):
+    """Base class for Github release assets pusher/fetcher"""
+
+    @abstractmethod
+    def __init__(self, release: "AGithubRelease", path: pathlib.Path) -> None:
+        raise NotImplementedError
```

Comment:
Could a decorator automatically raise this once we declare a method as abstract? I'm thinking of the comparable Envoy experience, where we mark virtual methods as `PURE`. Having to raise an error feels less declarative and more plumbing like on an abstract interface.

Reply:
using `NotImplementedError` is the ~standard way of declaring a method as an interface method (at least in zope, but i see it fairly commonly in other abstractmethod examples)

re it being raised - it could theoretically be raised if you were to do `super().__ini__()` - as (at least currently) implementers inherit from the interfaces

it could not be raised by omission - if you dont override this function in an implementer the class wont instantiate

this is kind of the core issue around whether to use `ABC.register(MyImpl)` or `MyImpl(ABC)`

Reply:
...the first way means you cant call `super()....` anyway because its not a descendant, but it also means that the implementer is not checked (at least by python) as implementing the interface

either way - interface or abstract class methods that dont implement anything, i think, should raise a `NotImplementedError` both because its more expressive of the actuality, and because its the right error to raise in the event that such a method did get hit

Reply:
so, thinking about this further

i guess we could add an `@interfacemethod` decorator based on `@abc.abstractmethod` that automatically raised an error

it would have the benefit of being more clearly not an abstract method and not callable

in terms of syntax it doesnt save us anything - the method still needs a body, so it would end up something like this

```python
@interfacemethod
def some_iface_method(self):
    raise NotImplementedError
```

one thing we could also do is add an `abstract.Interface` class which was like an `abstract.Abstraction` but only allowed interface methods, and autoraised etc if any were called - ie a "pure" interface

i guess if we did these things i would feel more comfortable using the abstract classes for abstract implementation because there would be more clarity as to what is callable etc

for this pr though, i reckon we should keep it simple. We can think about adding the above as use cases emerge